### ISSUE TITLE
Allow UnmarshalNew() in the ErrorDetail interface

### ISSUE THREAD

**Issue by ctsquires** [bufbuild/connect-go#328]
**Is your feature request related to a problem? Please describe.**

Quality of life improvement for parsing `ErrorDetail` as a client.

Currently parsing the slice of ErrorDetails into many possible underlying types feels a bit repetitive and we can't use a standard switch case on the type to figure out which error detail was returned. It would be nice if the ErrorDetail interface exposed the `UnmarshalNew()` functionality of the underlying `anypb.Any` type to allow using a traditional switch case.

This functionality is referenced in the anypb documentation https://pkg.go.dev/google.golang.org/protobuf/types/known/anypb#hdr-Unmarshaling_an_Any. 

Currently I believe clients need to iterate through every type and details, checking if the detail can successfully be unmarshalled to that type and then actioning if it successfully unmarshals.

eg. We currently have a few types that can be returned, and we construct an internal error model to hold these types if we receive them from a server, with the current setup we need to do something like below (we could iterate through the details and check each type, but both end up with similar amounts of code):

```go

func extractErrorDetails(connectErr *connect.Error) (MyErrorType, error) {
	var errorOpts []myErrorOption
	retryInfo := tryExtractRetryInfo(connectErr)
	if retryInfo != nil {
		errorOpts = append(errorOpts, retryInfoOption(retryInfo))
	}
	helpInfo := tryExtractHelpInfo(connectErr)
	if helpInfo != nil {
		errorOpts = append(errorOpts, helpInfoOption(retryInfo))
	}
	return newMyErrorType(errorOpts...), nil
}

func tryExtractRetryInfo(connectErr *connect.Error) *pb.RetryInfo {
	var retryInfo errdetails.RetryInfo
	for _, any := range connectErr.Details() {
		if err := any.UnmarshalTo(&retryInfo); err == nil {
			return &retryInfo
		}
	}
}

func tryExtractHelpInfo(connectErr *connect.Error) (*pb.HelpInfo, ok) {
	var helpInfo errdetails.HelpInfo
	for _, any := range connectErr.Details() {
		if err := any.UnmarshalTo(&helpInfo); err == nil {
			return &helpInfo
		}
	}
}
```

**Describe the solution you'd like**

It would be nice if the `ErrorDetail` interface exposed `UnmarshalNew()` 

```go
type ErrorDetail interface {
	proto.Message

	MessageName() protoreflect.FullName
	UnmarshalTo(proto.Message) error
	UnmarshalNew() (proto.Message, error) // new, anypb.Any already implements
}
```

this would allow clients to use a traditional switch case on `.(type)` to figure out the underlying message and action

```go
func extractErrorDetails(c *connect.Error) (MyErrorType, error) {
	var errorOpts []myErrorOption
	for _, detail := range connectErr.Details() {
		d, err := detail.UnmarshalNew()
		if err != nil {
			continue
		}
		switch ext := d.(type) {
		case *errdetails.HelpInfo:
			errorOpts = append(errorOpts, helpInfoOption(d))
		case *errdetails.RetryInfo:
			errorOpts = append(errorOpts, retryInfoOption(d))
		}
	}
	return newMyErrorType(errorOpts...), nil
}
```

**Describe alternatives you've considered**

* Converting to anypb.Any allows the full range of functionality and allows the switch case setup to work, but this felt like sidestepping the abstraction ErrorDetail.
* We can use the unmarshalTo method as shown above, but this ends up with a lot of boilerplate, especially as the number of possible type options increases.

**Comment by akshayjshah** [comment]
Great point, @ctsquires! We're re-working error details right now to make it easier to write proxies that translate between the gRPC protocols and the Connect protocol, and I think it'll be easy to address this as part of that change.

**Comment by akshayjshah** [comment]
Fixed in #329.
